src/exp/trainers/base.py

```diff
@@ -0,0 +1,108 @@
+from pamiq_core.torch import TorchTrainer
+from torch.utils.data import DataLoader
+from tqdm import tqdm
+
+from exp.aim_utils import get_global_run
+
+
+class ExperimentTrainer(TorchTrainer, ABC):
+    """Base trainer class for vision experiments.
+
+    This abstract base class provides common functionality for training
+    vision models in the PAMIQ framework. It handles training loops,
+    progress tracking, and integration with Aim for experiment tracking.
+    """
+
+    global_steps: int = 0
+    """Total number of training steps across all epochs."""
+
+    # Please override this parameters in constructor
+    max_epochs: int = 1
+    """Maximum number of epochs to train in a single training session."""
+
+    max_steps_every_train: int | None = None
+    """Maximum number of steps to train per session.
+
+    If set, training stops after this many steps.
+    """
+
+    trainer_name: str = "experiment"
+    """Name of the trainer used for Aim context logging."""
+
+    @property
+    def default_aim_context(self) -> AimObjectDict:
+        return {"trainer": self.trainer_name}
+
+    @abstractmethod
+    def create_dataloader(self) -> DataLoader[Any]:
+        """Create and return a DataLoader for training.
+
+        Subclasses must implement this method to provide training data.
+
+        Returns:
+            DataLoader configured for the specific training task.
+        """
+        pass
+
+    def batch_step(self, batch: Any, index: int) -> None:
+        """Process a single batch of data.
+
+        Subclasses should override this method to implement their training logic.
+
+        Args:
+            batch: A batch of data from the dataloader.
+            index: The index of the current batch within the epoch.
+        """
+        pass
+
+    @override
+    def setup(self) -> None:
+        super().setup()
+        self.current_epoch = 0
+        self.terminate_training = False
+
+        if (aim_run := get_global_run()) is None:
+            raise ValueError(
+                "No global Aim run found. Please initialize Aim before training."
+            )
+        self.aim_run = aim_run
+
+    @override
+    def train(self) -> None:
+        dataloader = self.create_dataloader()
+        start_global_step = self.global_steps
+        for epoch in tqdm(range(self.max_epochs), desc="Epoch"):
+            self.current_epoch = epoch
+            for idx, batch in enumerate(tqdm(dataloader, leave=False, desc="Batch")):
+                self.batch_step(batch, idx)
+                self.global_steps += 1
+
+                if self.max_steps_every_train is not None:
+                    if (
+                        self.global_steps - start_global_step
+                        >= self.max_steps_every_train
+                    ):
+                        self.terminate_training = True
+
+                if self.terminate_training:
+                    break
+            if self.terminate_training:
+                break
```

Comment:
`self.terminate_training`がいらない気がしてきた
```suggestion
                if self.max_steps_every_train is not None:
                    if (
                        self.global_steps - start_global_step
                        >= self.max_steps_every_train
                    ):
                        return
```

Reply:
修正しました。https://github.com/MLShukai/pamiq-vision-exp/pull/8/commits/883dfeaf5cae0f1cca6cbafa4d5a9516e97079c7